phd-tests/tests/src/cpuid.rs

```diff
@@ -0,0 +1,178 @@
+    cfg.cpuid(entries.clone());
+    let mut vm = ctx.spawn_vm(&cfg, None).await?;
+    vm.launch().await?;
+
+    let spec_get_response = vm.get_spec().await?;
+    let propolis_client::types::VersionedInstanceSpec::V0(spec) =
+        spec_get_response.spec;
+
+    let cpuid = spec.board.cpuid.expect("board should have explicit CPUID");
+    assert_eq!(cpuid.entries.len(), entries.len());
+    itertools::assert_equal(cpuid.entries, entries);
+}
+
+#[phd_testcase]
+async fn cpuid_boot_test(ctx: &Framework) {
+    use cpuid_utils::bits::*;
+
+    // This test verifies that plumbing a reasonably sensible-looking set of
+    // CPUID values into the guest produces a bootable guest.
+    //
+    // The definition of "reasonably sensible" is derived from the suggested
+    // virtual-Milan CPUID definitions in RFD 314. Those are in turn derived
+    // from reading AMD's manuals and bhyve's source code to figure out what
+    // host CPU features to hide from guests.
+    //
+    // To try to make this test at least somewhat host-agnostic, read all of the
+    // host's CPUID leaves, then filter to just a handful of leaves that
+    // advertise features to the guest (and that Linux guests will check for
+    // during boot).
+    let mut host_cpuid = cpuid_utils::host_query_all().0;
+    info!(?host_cpuid, "read host CPUID");
+    let leaf_0_values = *host_cpuid
+        .get(&CpuidIdent::leaf(0))
+        .expect("host CPUID leaf 0 should always be present");
+
+    // Linux guests expect to see at least a couple of leaves in the extended
+    // CPUID range. These have vendor-specific meanings. This test only encodes
+    // AMD's definitions, so skip the test if leaf 0 reports any other vendor.
+    let Ok(CpuidVendor::Amd) = CpuidVendor::try_from(leaf_0_values) else {
+        phd_skip!("cpuid_boot_test can only run on AMD hosts");
+    };
+
+    // Report up through leaf 7 to get the extended feature flags that are
+    // defined there.
+    if leaf_0_values.eax < 7 {
+        phd_skip!(
+            "cpuid_boot_test requires the host to report at least leaf 7"
+        );
+    }
+
+    // Keep only standard leaves up to 7 and the first two extended leaves.
+    // Extended leaves 2 through 4 will be overwritten with a fake brand string
+    // (see below).
+    host_cpuid.retain(|leaf, _values| {
+        leaf.leaf <= 0x7
+            || (leaf.leaf >= 0x8000_0000 && leaf.leaf <= 0x8000_0001)
+    });
+
+    // Report that leaf 7 is the last available standard leaf.
+    host_cpuid.get_mut(&CpuidIdent::leaf(0)).unwrap().eax = 7;
+
+    // Mask off bits that a minimum-viable Oxide platform doesn't support or
+    // can't (or won't) expose to guests. See RFD 314 for further discussion of
+    // how these masks were chosen.
+    //
+    // These are masks (and not assignments) so that, if the host processor
+    // doesn't support a feature contained in an `ALL_FLAGS` value, the
```

Comment:
this (much nicer) comment lost its tail end. `the feature is not incorrectly advertised`?

Reply:
argh

Reply:
Fixed in c560647.